client/js/s3/request-signer.js

```diff
@@ -302,6 +307,32 @@ qq.s3.RequestSigner = function(o) {
             errorMessage = "Received an empty or invalid response from the server!";
         }
 
+        response = options.onSigningRequestComplete(response, !isError, xhrOrXdr, tempCredentials);
+
+        if (tempCredentials) {
+            if (Object.keys) {
+                Object.keys(tempCredentials).length === 0 && (tempCredentials = null);
+            }
+            else {
+                /** this is for pre-ECMA 5 compatability */
+                for (i in tempCredentials) {
+                    if (tempCredentials.hasOwnProperty(i)) {
+                        tempCredentials = null;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (tempCredentials) {
+            options.log("Temporary Amazon credentials found after onSigningRequestComplete event callback");
+            credentials = tempCredentials;
+        } else {
+            credentials = credentialsProvider.get();
+        }
+
+        isError = isError || (!response) && Boolean(errorMessage = "onSigningRequestComplete event callback returned false");
```

Comment:
this and other areas have been updated to reflect making the code less dense, easier to follow.

this part here with the `isError` detection was a little convoluted, my apologies.  content of `errorMessage` was not being used to determine error state, rather it was `!response` which if true would ensure that `errorMessage` gets assigned properly (and after being cast to boolean `errorMessage === !response === true`).

But I have made this more clear and better now, nevermind what it was before.